{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create a manifest file\n",
    "\n",
    "The *manifest.txt* file created here is intended to be used with [gdc-client](https://gdc.cancer.gov/access-data/gdc-data-transfer-tool)\n",
    "\n",
    "Please refer to [gdc API](https://docs.gdc.cancer.gov/API/Users_Guide/Getting_Started/) for reference about query done here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Author: Filippo_Valle\n",
      "\n",
      "Last updated: Tue Feb 08 2022\n",
      "\n",
      "Python implementation: CPython\n",
      "Python version       : 3.8.12\n",
      "IPython version      : 7.22.0\n",
      "\n",
      "pandas  : 1.2.4\n",
      "requests: 2.25.1\n",
      "json    : 2.0.9\n",
      "\n",
      "Compiler    : Clang 11.1.0 \n",
      "OS          : Darwin\n",
      "Release     : 21.2.0\n",
      "Machine     : x86_64\n",
      "Processor   : i386\n",
      "CPU cores   : 8\n",
      "Architecture: 64bit\n",
      "\n",
      "Git hash: 92731dbe84ae34b2ef487d4bd4db5248c4343735\n",
      "\n",
      "Git repo: https://github.com/SnigdhaJamwal/BoltBio.git\n",
      "\n",
      "Git branch: kushan\n",
      "\n",
      "Watermark: 2.3.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%load_ext watermark\n",
    "%watermark -v -m  -u -n -p pandas,requests,json -a Filippo_Valle -g -r -b -w"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "\n",
    "\n",
    "PATH = os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use file endopint\n",
    "files_endpt = \"https://api.gdc.cancer.gov/files\"\n",
    "\n",
    "# The 'fields' parameter is passed as a comma-separated string of single names.\n",
    "fields = [\n",
    "    \"file_name\",\n",
    "    #\"analysis.workflow_type\",\n",
    "    #\"experimental_strategy\",\n",
    "    \"cases.project.primary_site\",\n",
    "    \"cases.project.disease_type\",\n",
    "    \"cases.project.project_id\",\n",
    "    \"cases.submitter_id\",\n",
    "    \"cases.samples.portions.analytes.aliquots.submitter_id\",\n",
    "    \"cases.demographic.vital_status\",\n",
    "    \"cases.demographic.days_to_birth\",\n",
    "    \"cases.demographic.days_to_death\",\n",
    "    \"cases.demographic.gender\",\n",
    "    \"cases.diagnoses.primary_diagnosis\",\n",
    "    \"cases.diagnoses.tumor_stage\",\n",
    "    \"cases.diagnoses.tumor_grade\",\n",
    "    \"cases.diagnoses.last_known_disease_status\",\n",
    "    \"cases.diagnoses.progression_or_recurrence\"\n",
    "    \"cases.diagnoses.days_to_last_follow_up\",\n",
    "    \"cases.diagnoses.age_at_diagnosis\",\n",
    "    \"cases.diagnoses.morphology\",\n",
    "    \"cases.diagnoses.tissue_or_organ_of_origin\",\n",
    "    \"cases.exposures.years_smoked\",\n",
    "    \"cases.exposures.cigarettes_per_day\",\n",
    "    \"cases.samples.longest_dimension\",\n",
    "    \"samples.portions.analytes.aliquots.submitter_id\",\n",
    "]\n",
    "\n",
    "fields = ','.join(fields)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter gene expression data obtained with *RNA-Seq HTSeq* in txt format either from **BRCA** or **LUAD** and **LUSC**\n",
    "\n",
    "Don't forget to select the correct filter for *cases.project.project_id*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tissue project filters\n",
    "filters = {\n",
    "    \"op\": \"and\",\n",
    "    \"content\":[\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.data_type\",\n",
    "            \"value\": [\"Gene Expression Quantification\"]\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.experimental_strategy\",\n",
    "            \"value\": [\"RNA-Seq\"]\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "               \"field\": \"files.analysis.workflow_type\",\n",
    "                \"value\": [\"HTSeq - FPKM\"]\n",
    "                }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.data_format\",\n",
    "            \"value\": [\"TXT\"]\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"cases.project.project_id\",\n",
    "            \"value\": ['TCGA-LUAD',] \n",
    "            }\n",
    "        }\n",
    "        \n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A POST is used, so the filter parameters can be passed directly as a Dict object.\n",
    "##remove manifest to have file fields\n",
    "params = {\n",
    "   \"return_type\": \"manifest\",\n",
    "    \"filters\": filters,\n",
    "    \"fields\": fields,\n",
    "    \"format\": \"TSV\",\n",
    "    \"size\": \"15000\"\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The parameters are passed to 'json' rather than 'params' in this case\n",
    "response = requests.post(files_endpt, headers = {\"Content-Type\": \"application/json\"}, json = params)\n",
    "\n",
    "#print(response.content.decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f\"{PATH}/manifest.txt\",\"w\") as manifest:\n",
    "    manifest.write(response.content.decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Files\n",
    "\n",
    "Write a file with all metadata, useful in next analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\n",
    "    \"filters\": json.dumps(filters),\n",
    "    \"fields\": \"primary_site,cases.project.project_id,file_name,cases.demographic.vital_status,cases.demographic.gender,cases.diagnoses.age_at_diagnosis,cases.diagnoses.days_to_last_follow_up,cases.demographic.days_to_death,cases.demographic.days_to_birth,cases.submitter_id,samples.portions.analytes.aliquots.submitter_id,cases.diagnoses.last_known_disease_status,cases.diagnoses.tumor_stage,cases.exposures.years_smoked,cases.exposures.cigarettes_per_day,cases.samples.portions.analytes.aliquots.submitter_id\",\n",
    "    \"format\": \"TSV\",\n",
    "    \"size\": \"15000\"\n",
    "    }\n",
    "response = requests.post(files_endpt, headers = {\"Content-Type\": \"application/json\"}, json = params)\n",
    "with open(f\"{PATH}/files.txt\",\"w\") as files:\n",
    "    files.write(response.content.decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cases.0.demographic.days_to_birth</th>\n",
       "      <th>cases.0.demographic.days_to_death</th>\n",
       "      <th>cases.0.demographic.gender</th>\n",
       "      <th>cases.0.demographic.vital_status</th>\n",
       "      <th>cases.0.diagnoses.0.age_at_diagnosis</th>\n",
       "      <th>cases.0.diagnoses.0.days_to_last_follow_up</th>\n",
       "      <th>cases.0.diagnoses.0.last_known_disease_status</th>\n",
       "      <th>cases.0.exposures.0.cigarettes_per_day</th>\n",
       "      <th>cases.0.exposures.0.years_smoked</th>\n",
       "      <th>cases.0.project.project_id</th>\n",
       "      <th>cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id</th>\n",
       "      <th>cases.0.submitter_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>file_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12fff07f-74d6-4ee0-9f4f-36a4d194d24f.FPKM.txt.gz</th>\n",
       "      <td>-25069.0</td>\n",
       "      <td>161.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Dead</td>\n",
       "      <td>25069.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not reported</td>\n",
       "      <td>1.369863</td>\n",
       "      <td>15.0</td>\n",
       "      <td>TCGA-LUAD</td>\n",
       "      <td>TCGA-MP-A4T8-01A-11R-A24X-07</td>\n",
       "      <td>TCGA-MP-A4T8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>05d2b1e9-22a4-4369-b622-63846d672875.FPKM.txt.gz</th>\n",
       "      <td>-27993.0</td>\n",
       "      <td>1790.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Dead</td>\n",
       "      <td>27993.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not reported</td>\n",
       "      <td>2.739726</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-LUAD</td>\n",
       "      <td>TCGA-MP-A4T6-01A-32R-A262-07</td>\n",
       "      <td>TCGA-MP-A4T6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c8fb3217-2cc9-47c0-84b6-b666c07624e6.FPKM.txt.gz</th>\n",
       "      <td>-18762.0</td>\n",
       "      <td>896.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Dead</td>\n",
       "      <td>18762.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not reported</td>\n",
       "      <td>3.041096</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-LUAD</td>\n",
       "      <td>TCGA-49-4510-01A-01R-1206-07</td>\n",
       "      <td>TCGA-49-4510</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  cases.0.demographic.days_to_birth  \\\n",
       "file_name                                                                             \n",
       "12fff07f-74d6-4ee0-9f4f-36a4d194d24f.FPKM.txt.gz                           -25069.0   \n",
       "05d2b1e9-22a4-4369-b622-63846d672875.FPKM.txt.gz                           -27993.0   \n",
       "c8fb3217-2cc9-47c0-84b6-b666c07624e6.FPKM.txt.gz                           -18762.0   \n",
       "\n",
       "                                                  cases.0.demographic.days_to_death  \\\n",
       "file_name                                                                             \n",
       "12fff07f-74d6-4ee0-9f4f-36a4d194d24f.FPKM.txt.gz                              161.0   \n",
       "05d2b1e9-22a4-4369-b622-63846d672875.FPKM.txt.gz                             1790.0   \n",
       "c8fb3217-2cc9-47c0-84b6-b666c07624e6.FPKM.txt.gz                              896.0   \n",
       "\n",
       "                                                 cases.0.demographic.gender  \\\n",
       "file_name                                                                     \n",
       "12fff07f-74d6-4ee0-9f4f-36a4d194d24f.FPKM.txt.gz                       male   \n",
       "05d2b1e9-22a4-4369-b622-63846d672875.FPKM.txt.gz                     female   \n",
       "c8fb3217-2cc9-47c0-84b6-b666c07624e6.FPKM.txt.gz                     female   \n",
       "\n",
       "                                                 cases.0.demographic.vital_status  \\\n",
       "file_name                                                                           \n",
       "12fff07f-74d6-4ee0-9f4f-36a4d194d24f.FPKM.txt.gz                             Dead   \n",
       "05d2b1e9-22a4-4369-b622-63846d672875.FPKM.txt.gz                             Dead   \n",
       "c8fb3217-2cc9-47c0-84b6-b666c07624e6.FPKM.txt.gz                             Dead   \n",
       "\n",
       "                                                  cases.0.diagnoses.0.age_at_diagnosis  \\\n",
       "file_name                                                                                \n",
       "12fff07f-74d6-4ee0-9f4f-36a4d194d24f.FPKM.txt.gz                               25069.0   \n",
       "05d2b1e9-22a4-4369-b622-63846d672875.FPKM.txt.gz                               27993.0   \n",
       "c8fb3217-2cc9-47c0-84b6-b666c07624e6.FPKM.txt.gz                               18762.0   \n",
       "\n",
       "                                                  cases.0.diagnoses.0.days_to_last_follow_up  \\\n",
       "file_name                                                                                      \n",
       "12fff07f-74d6-4ee0-9f4f-36a4d194d24f.FPKM.txt.gz                                         NaN   \n",
       "05d2b1e9-22a4-4369-b622-63846d672875.FPKM.txt.gz                                         NaN   \n",
       "c8fb3217-2cc9-47c0-84b6-b666c07624e6.FPKM.txt.gz                                         NaN   \n",
       "\n",
       "                                                 cases.0.diagnoses.0.last_known_disease_status  \\\n",
       "file_name                                                                                        \n",
       "12fff07f-74d6-4ee0-9f4f-36a4d194d24f.FPKM.txt.gz                                  not reported   \n",
       "05d2b1e9-22a4-4369-b622-63846d672875.FPKM.txt.gz                                  not reported   \n",
       "c8fb3217-2cc9-47c0-84b6-b666c07624e6.FPKM.txt.gz                                  not reported   \n",
       "\n",
       "                                                  cases.0.exposures.0.cigarettes_per_day  \\\n",
       "file_name                                                                                  \n",
       "12fff07f-74d6-4ee0-9f4f-36a4d194d24f.FPKM.txt.gz                                1.369863   \n",
       "05d2b1e9-22a4-4369-b622-63846d672875.FPKM.txt.gz                                2.739726   \n",
       "c8fb3217-2cc9-47c0-84b6-b666c07624e6.FPKM.txt.gz                                3.041096   \n",
       "\n",
       "                                                  cases.0.exposures.0.years_smoked  \\\n",
       "file_name                                                                            \n",
       "12fff07f-74d6-4ee0-9f4f-36a4d194d24f.FPKM.txt.gz                              15.0   \n",
       "05d2b1e9-22a4-4369-b622-63846d672875.FPKM.txt.gz                               NaN   \n",
       "c8fb3217-2cc9-47c0-84b6-b666c07624e6.FPKM.txt.gz                               NaN   \n",
       "\n",
       "                                                 cases.0.project.project_id  \\\n",
       "file_name                                                                     \n",
       "12fff07f-74d6-4ee0-9f4f-36a4d194d24f.FPKM.txt.gz                  TCGA-LUAD   \n",
       "05d2b1e9-22a4-4369-b622-63846d672875.FPKM.txt.gz                  TCGA-LUAD   \n",
       "c8fb3217-2cc9-47c0-84b6-b666c07624e6.FPKM.txt.gz                  TCGA-LUAD   \n",
       "\n",
       "                                                 cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id  \\\n",
       "file_name                                                                                                          \n",
       "12fff07f-74d6-4ee0-9f4f-36a4d194d24f.FPKM.txt.gz                       TCGA-MP-A4T8-01A-11R-A24X-07                \n",
       "05d2b1e9-22a4-4369-b622-63846d672875.FPKM.txt.gz                       TCGA-MP-A4T6-01A-32R-A262-07                \n",
       "c8fb3217-2cc9-47c0-84b6-b666c07624e6.FPKM.txt.gz                       TCGA-49-4510-01A-01R-1206-07                \n",
       "\n",
       "                                                 cases.0.submitter_id  \n",
       "file_name                                                              \n",
       "12fff07f-74d6-4ee0-9f4f-36a4d194d24f.FPKM.txt.gz         TCGA-MP-A4T8  \n",
       "05d2b1e9-22a4-4369-b622-63846d672875.FPKM.txt.gz         TCGA-MP-A4T6  \n",
       "c8fb3217-2cc9-47c0-84b6-b666c07624e6.FPKM.txt.gz         TCGA-49-4510  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_files = pd.read_csv(f\"{PATH}/files.txt\", sep='\\t')\n",
    "df_files.drop(\"id\", axis=1, inplace=True)\n",
    "df_files.set_index(\"file_name\", inplace=True)\n",
    "df_files.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['cases.0.demographic.days_to_birth',\n",
       "       'cases.0.demographic.days_to_death', 'cases.0.demographic.gender',\n",
       "       'cases.0.demographic.vital_status',\n",
       "       'cases.0.diagnoses.0.age_at_diagnosis',\n",
       "       'cases.0.diagnoses.0.days_to_last_follow_up',\n",
       "       'cases.0.diagnoses.0.last_known_disease_status',\n",
       "       'cases.0.exposures.0.cigarettes_per_day',\n",
       "       'cases.0.exposures.0.years_smoked', 'cases.0.project.project_id',\n",
       "       'cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id',\n",
       "       'cases.0.submitter_id'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_files.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save file to dat\n",
    "df_files.to_csv(f\"{PATH}/files_tcga.dat\", header=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
